File: OnlyCueTests/MTCScheduleGoldenVectorTests.swift
import XCTest
@testable import OnlyCue

// The generator + drift guard for `golden/mtc-schedule-v1.json`. The contract
// model and the case matrix live in `MTCScheduleGolden.swift`.

extension MTCScheduleGolden {

    // MARK: - Case builders

    private static func baseInput(_ schedule: MTCSchedule) -> MTCScheduleGoldenVector.Input {
        MTCScheduleGoldenVector.Input(
            hours: schedule.startTimecode.hours,
            minutes: schedule.startTimecode.minutes,
            seconds: schedule.startTimecode.seconds,
            frames: schedule.startTimecode.frames,
            anchorHostTime: schedule.anchorHostTime,
            ticksPerSecond: GoldenDouble(schedule.ticksPerSecond)
        )
    }

    private static func cadenceCase(
        _ rate: SMPTEFramerate,
        ticksPerSecond: Double
    ) throws -> MTCScheduleGoldenVector.Case {
        let schedule = try schedule(rate, ticksPerSecond)
        return MTCScheduleGoldenVector.Case(
            label: "cadence/\(rate.rawValue)/\(label(ticksPerSecond))",
            op: "cadence",
            rate: rate.rawValue,
            input: baseInput(schedule),
            expect: MTCScheduleGoldenVector.Expect(
                ticksPerQuarterFrame: GoldenDouble(schedule.ticksPerQuarterFrame)
            )
        )
    }

    private static func sequenceCase(_ rate: SMPTEFramerate, index: Int) throws -> MTCScheduleGoldenVector.Case {
        let schedule = try schedule(rate, 1_000_000_000)
        var input = baseInput(schedule)
        input.sequenceIndex = index
        return MTCScheduleGoldenVector.Case(
            label: "sequenceTimecode/\(rate.rawValue)/\(schedule.startTimecode.displayString)/seq\(index)",
            op: "sequenceTimecode",
            rate: rate.rawValue,
            input: input,
            expect: MTCScheduleGoldenVector.Expect(
                timecode: schedule.timecode(forSequence: index).displayString
            )
        )
    }

    private static func quarterFrameCase(
        _ rate: SMPTEFramerate,
        ticksPerSecond: Double,
        index: Int
    ) throws -> MTCScheduleGoldenVector.Case {
        let schedule = try schedule(rate, ticksPerSecond)
        var input = baseInput(schedule)
        input.quarterFrameIndex = index
        return MTCScheduleGoldenVector.Case(
            label: "quarterFrame/\(rate.rawValue)/\(label(ticksPerSecond))/index\(index)",
            op: "quarterFrame",
            rate: rate.rawValue,
            input: input,
            expect: MTCScheduleGoldenVector.Expect(
                byte: Int(schedule.byte(forQuarterFrame: index)),
                timestamp: schedule.timestamp(forQuarterFrame: index)
            )
        )
    }

    private static func windowCase(_ spec: WindowCase) throws -> MTCScheduleGoldenVector.Case {
        let schedule = try schedule(.fps24, 1_000_000_000)
        let from = tick(schedule, spec.from)
        let until = tick(schedule, spec.until)
        var input = baseInput(schedule)
        input.from = from
        input.until = until
        return MTCScheduleGoldenVector.Case(
            label: "batch/24/1e9/\(spec.label)",
            op: "batch",
            rate: SMPTEFramerate.fps24.rawValue,
            input: input,
            expect: MTCScheduleGoldenVector.Expect(
                messages: encode(schedule.batch(from: from, until: until))
            )
        )
    }

    private static func chainCase(_ spec: ChainCase) throws -> MTCScheduleGoldenVector.Case {
        let schedule = try schedule(spec.rate, spec.ticksPerSecond)
        let boundaries = spec.boundaries.map { tick(schedule, $0) }
        let label = "batchChain/\(spec.rate.rawValue)/\(label(spec.ticksPerSecond))/\(spec.note)"
        guard zip(boundaries, boundaries.dropFirst()).allSatisfy({ $0 < $1 }) else {
            throw MTCScheduleGoldenError.boundariesNotIncreasing(label: label)
        }
        var input = baseInput(schedule)
        input.boundaries = boundaries
        let windows = zip(boundaries, boundaries.dropFirst()).map {
            encode(schedule.batch(from: $0, until: $1))
        }
        return MTCScheduleGoldenVector.Case(
            label: label,
            op: "batchChain",
            rate: spec.rate.rawValue,
            input: input,
            expect: MTCScheduleGoldenVector.Expect(
                windows: windows,
                combined: encode(schedule.batch(from: boundaries[0], until: boundaries[boundaries.count - 1]))
            )
        )
    }

    // MARK: - Assembly

    static func make() throws -> MTCScheduleGoldenVector {
        var cases = try allRates.flatMap { rate in
            try clocks.map { try cadenceCase(rate, ticksPerSecond: $0) }
        }
        cases += try allRates.flatMap { rate in
            try sequenceIndices.map { try sequenceCase(rate, index: $0) }
        }
        cases += try messageClocks.flatMap { clock in
            try quarterFrameIndices.map {
                try quarterFrameCase(clock.rate, ticksPerSecond: clock.ticksPerSecond, index: $0)
            }
        }
        cases += try windowCases.map(windowCase)
        cases += try chainCases.map(chainCase)
        return MTCScheduleGoldenVector(
            contract: "mtc-schedule",
            version: 1,
            note: "macOS-generated golden vectors for the OnlyCue MTC scheduling layer "
                + "(MTCSchedule). The C# OnlyCue.Core re-implementation must reproduce every "
                + "case exactly. batchChain is the load-bearing op: concat(windows) must equal "
                + "combined, so a message dropped at a tiled seam and one duplicated there fail "
                + "differently. hostTicksPerSecond() is out of contract — it is Mach-specific. "
                + "To regenerate after a deliberate change, delete golden/mtc-schedule-v1.json "
                + "and re-run the suite.",
            cases: cases
        )
    }

    /// Deterministic encoding so the committed file is stable and diff-friendly.
    static func encoded(_ vector: MTCScheduleGoldenVector) throws -> Data {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes]
        return try encoder.encode(vector)
    }
}

// MARK: - Tests

final class MTCScheduleGoldenVectorTests: XCTestCase {

    private func goldenURL() throws -> URL {
        try repoRoot().appendingPathComponent("golden", isDirectory: true)
            .appendingPathComponent("mtc-schedule-v1.json")
    }

    /// Independent pins that do **not** go through the generator. Hand-computed
    /// from the cadence rather than read off the implementation.
    func test_knownScheduleValues() throws {
        let schedule = try MTCScheduleGolden.schedule(.fps25, 1_000_000_000)
        // 25 fps × 4 = 100 messages per second at nanosecond ticks.
        XCTAssertEqual(schedule.ticksPerQuarterFrame, 10_000_000)
        XCTAssertEqual(schedule.timestamp(forQuarterFrame: 0), MTCScheduleGolden.anchor)
        XCTAssertEqual(schedule.timestamp(forQuarterFrame: 8), MTCScheduleGolden.anchor + 80_000_000)
        // Negative indices clamp rather than running backwards past the anchor.
        XCTAssertEqual(schedule.timestamp(forQuarterFrame: -5), MTCScheduleGolden.anchor)
        XCTAssertEqual(schedule.byte(forQuarterFrame: -5), schedule.byte(forQuarterFrame: 0))
        // Eight messages carry one value and advance it by two frames.
        XCTAssertEqual(schedule.timecode(forSequence: 0), schedule.startTimecode)
        XCTAssertEqual(
            schedule.timecode(forSequence: 1).frameCount,
            schedule.startTimecode.frameCount + 2
        )
    }

    /// The quantise in `quarterFrameIndex(atOrAfter:)` is load-bearing, not
    /// padding. Reproduces what a port without it would do and counts the
    /// damage, so the number in the contract's header comment is an assertion
    /// rather than a claim: at nanosecond ticks a tiled chain would drop more
    /// than a third of all messages at 24 fps, and none at 25 fps, where the
    /// quarter-frame period is a whole number of ticks.
    func test_theQuantiseIsLoadBearing_atNanosecondTicks() throws {
        func dropped(rate: SMPTEFramerate) throws -> Int {
            let schedule = try MTCScheduleGolden.schedule(rate, 1_000_000_000)
            let period = schedule.ticksPerQuarterFrame
            return (0..<200).filter { index in
                let offset = Double(schedule.timestamp(forQuarterFrame: index)) - Double(MTCScheduleGolden.anchor)
                guard offset > 0 else { return false }
                // No quantise: `ceil` of a value a hair above the integer index
                // lands on the *next* index, so this message never gets emitted.
                return Int((offset / period).rounded(.up)) != index
            }.count
        }
        XCTAssertEqual(try dropped(rate: .fps24), 74)
        XCTAssertEqual(try dropped(rate: .fps30), 66)
        XCTAssertEqual(try dropped(rate: .fps25), 0, "the period is exactly 10 000 000 ticks here")
    }

    /// Tiling is the property the windowing exists to provide: passing the
    /// previous call's `until` as the next call's `from` must yield every
    /// message exactly once. Asserted over the committed chains directly, so a
    /// chain whose vector was regenerated from a broken implementation still
    /// fails here.
    func test_tiledWindows_yieldEveryMessageExactlyOnce() throws {
        for spec in MTCScheduleGolden.chainCases {
            let schedule = try MTCScheduleGolden.schedule(spec.rate, spec.ticksPerSecond)
            let boundaries = spec.boundaries.map { MTCScheduleGolden.tick(schedule, $0) }
            let windows = zip(boundaries, boundaries.dropFirst()).map {
                schedule.batch(from: $0, until: $1)
            }
            let combined = schedule.batch(from: boundaries[0], until: boundaries[boundaries.count - 1])
            let label = "\(spec.rate.rawValue)/\(MTCScheduleGolden.label(spec.ticksPerSecond))/\(spec.note)"
            XCTAssertEqual(windows.flatMap { $0 }, combined, label)
            XCTAssertFalse(combined.isEmpty, "\(label): a chain that yields nothing pins nothing")
            XCTAssertEqual(
                Set(combined.map(\.timestamp)).count,
                combined.count,
                "\(label): no message may appear twice"
            )
        }
    }

    /// The coarse clock has to *be* coarse enough to make every odd index a
    /// rounding tie, or the only case separating half-away-from-zero from
    /// banker's rounding inside `timestamp(forQuarterFrame:)` goes quiet without
    /// failing.
    func test_theCoarseClockCaseIsActuallyATie() throws {
        let schedule = try MTCScheduleGolden.schedule(.fps25, 50)
        XCTAssertEqual(schedule.ticksPerQuarterFrame, 0.5)
        // Half away from zero: index 1 is at 0.5 ticks → 1, not the 0 banker's
        // rounding would give.
        XCTAssertEqual(schedule.timestamp(forQuarterFrame: 1), MTCScheduleGolden.anchor + 1)
        XCTAssertEqual(schedule.timestamp(forQuarterFrame: 5), MTCScheduleGolden.anchor + 3)
        XCTAssertTrue(MTCScheduleGolden.messageClocks.contains { $0.ticksPerSecond == 50 })
    }

    /// Drift guard + bootstrap. When the committed file is present it must equal
    /// the current Swift output byte-for-byte (that's the contract the C# core
    /// verifies against). When it is missing — first run, or after a deliberate
    /// change where the dev deleted it to regenerate — it is written and the test
    /// fails, so the new contract is committed and reviewed rather than silently
    /// accepted in CI.
    func test_committedVectors_matchCurrentSwiftOutput() throws {
        let url = try goldenURL()
        let generated = try MTCScheduleGolden.encoded(MTCScheduleGolden.make())

        guard FileManager.default.fileExists(atPath: url.path) else {
            try FileManager.default.createDirectory(
                at: url.deletingLastPathComponent(), withIntermediateDirectories: true
            )
            try generated.write(to: url)
            return XCTFail("golden/mtc-schedule-v1.json was missing — generated it from the Swift "
                + "MTCSchedule implementation. Commit the file and re-run. (To regenerate after a "
                + "deliberate change, delete the file first.)")
        }

        let committed = try JSONDecoder().decode(
            MTCScheduleGoldenVector.self, from: Data(contentsOf: url)
        )
        XCTAssertEqual(
            committed,
            try MTCScheduleGolden.make(),
            "golden/mtc-schedule-v1.json drifted from the Swift implementation. If the change "
                + "was intentional, delete the file and re-run to regenerate, review the "
                + "diff, and update the C# OnlyCue.Core to match."
        )
    }
}

File: windows/OnlyCue.Core.Tests/MtcScheduleGoldenVectorTests.cs
using OnlyCue.Core.Midi;
using Xunit;

namespace OnlyCue.Core.Tests;

/// <summary>
/// The Windows half of the MTC scheduling contract (epic #728, M3 slice 2).
/// macOS emits <c>golden/mtc-schedule-v1.json</c> from the Swift
/// <c>MTCSchedule</c>; this suite asserts the C# re-implementation reproduces
/// every case exactly.
/// </summary>
/// <remarks>
/// <c>batchChain</c> is the load-bearing op: the windows are tiled, so
/// <c>concat(windows)</c> must equal <c>combined</c>. A message dropped at a seam
/// and one duplicated there fail differently, which is what makes the quantise in
/// <c>QuarterFrameIndexAtOrAfter</c> testable at all.
/// </remarks>
public class MtcScheduleGoldenVectorTests
{
    private static readonly MtcScheduleVector Vector = MtcScheduleVector.Load();

    public static TheoryData<string> CaseLabels
    {
        get
        {
            var data = new TheoryData<string>();
            foreach (var label in Vector.Cases.Select(c => c.Label))
            {
                data.Add(label);
            }

            return data;
        }
    }

    [Fact]
    public void GoldenFile_IsTheExpectedContract()
    {
        Assert.Equal("mtc-schedule", Vector.Contract);
        Assert.Equal(1, Vector.Version);
        Assert.NotEmpty(Vector.Cases);
    }

    [Theory]
    [MemberData(nameof(CaseLabels))]
    public void CSharpCore_ReproducesGoldenCase(string label)
    {
        var goldenCase = Vector.Cases.Single(c => c.Label == label);
        var schedule = ScheduleFrom(goldenCase);

        switch (goldenCase.Op)
        {
            case "cadence":
                AssertDouble(
                    goldenCase.Expect.TicksPerQuarterFrame,
                    schedule.TicksPerQuarterFrame,
                    label,
                    "ticksPerQuarterFrame");
                break;
            case "sequenceTimecode":
                Assert.Equal(
                    goldenCase.Expect.Timecode,
                    schedule.TimecodeForSequence(Required(goldenCase.Input.SequenceIndex, label, "sequenceIndex"))
                        .DisplayString);
                break;
            case "quarterFrame":
                AssertQuarterFrame(goldenCase, schedule);
                break;
            case "batch":
                Assert.Equal(
                    goldenCase.Expect.Messages,
                    Encode(schedule.Batch(
                        Required(goldenCase.Input.From, label, "from"),
                        Required(goldenCase.Input.Until, label, "until"))));
                break;
            case "batchChain":
                AssertChain(goldenCase, schedule);
                break;
            default:
                throw new InvalidDataException($"unknown golden op '{goldenCase.Op}'");
        }
    }

    private static void AssertQuarterFrame(MtcScheduleCase goldenCase, MtcSchedule schedule)
    {
        var index = Required(goldenCase.Input.QuarterFrameIndex, goldenCase.Label, "quarterFrameIndex");
        Assert.Equal(goldenCase.Expect.Byte, schedule.ByteForQuarterFrame(index));
        Assert.Equal(goldenCase.Expect.Timestamp, schedule.TimestampForQuarterFrame(index));
    }

    private static void AssertChain(MtcScheduleCase goldenCase, MtcSchedule schedule)
    {
        var label = goldenCase.Label;
        var boundaries = Required(goldenCase.Input.Boundaries, label, "boundaries");
        var windows = new List<IReadOnlyList<IReadOnlyList<ulong>>>(
            boundaries.Zip(boundaries.Skip(1), (from, until) => Encode(schedule.Batch(from, until))));
        var combined = Encode(schedule.Batch(boundaries[0], boundaries[^1]));

        Assert.Equal(goldenCase.Expect.Windows, windows);
        Assert.Equal(goldenCase.Expect.Combined, combined);

        // The tiling property itself, asserted against this core's own output
        // rather than against the vector — so a chain whose vector was somehow
        // regenerated from a broken implementation still fails here.
        Assert.Equal(combined, windows.SelectMany(window => window).ToList());
        Assert.NotEmpty(combined);
        Assert.Equal(combined.Count, combined.Select(message => message[1]).Distinct().Count());
    }

    private static MtcSchedule ScheduleFrom(MtcScheduleCase goldenCase)
    {
        var label = goldenCase.Label;
        var rate = SmpteFramerateExtensions.FromRawValue(goldenCase.Rate);
        var timecode = Timecode.Create(
                           Required(goldenCase.Input.Hours, label, "hours"),
                           Required(goldenCase.Input.Minutes, label, "minutes"),
                           Required(goldenCase.Input.Seconds, label, "seconds"),
                           Required(goldenCase.Input.Frames, label, "frames"),
                           rate)
                       ?? throw new InvalidDataException($"{label} does not name a valid timecode");

        return new MtcSchedule(
            timecode,
            Required(goldenCase.Input.AnchorHostTime, label, "anchorHostTime"),
            GoldenDouble.Parse(Required(goldenCase.Input.TicksPerSecond, label, "ticksPerSecond")));
    }

    /// <summary><c>[[byte, timestamp], …]</c> — the byte is already pinned by
    /// <c>golden/mtc-wire-v1.json</c>, so a mismatch here localises to the
    /// schedule.</summary>
    private static IReadOnlyList<IReadOnlyList<ulong>> Encode(IReadOnlyList<MtcSchedule.Message> messages) =>
        messages.Select(IReadOnlyList<ulong> (message) => new ulong[] { message.Byte, message.Timestamp }).ToList();

    private static int Required(int? value, string label, string field) =>
        value ?? throw new InvalidDataException($"{label} has no {field}");

    private static ulong Required(ulong? value, string label, string field) =>
        value ?? throw new InvalidDataException($"{label} has no {field}");

    private static T Required<T>(T? value, string label, string field)
        where T : class =>
        value ?? throw new InvalidDataException($"{label} has no {field}");

    private static void AssertDouble(string? expected, double actual, string label, string field)
    {
        var wanted = GoldenDouble.Parse(Required(expected, label, field));
        Assert.True(
            GoldenDouble.BitwiseEquals(wanted, actual),
            $"{label}: {field} expected {GoldenDouble.Describe(wanted)}, got {GoldenDouble.Describe(actual)}");
    }

    // MARK: - Independent pins (not read off the vectors)

    /// <summary>Mirrors <c>test_knownScheduleValues</c> on the Swift side.
    /// Hand-computed from the cadence, so a wrong C# implementation fails here
    /// even if the golden file were ever regenerated from a wrong source.</summary>
    [Fact]
    public void KnownCadence_AndTheTwoFrameConvention()
    {
        const ulong anchor = 1000;
        var start = Timecode.Create(0, 0, 59, 20, SmpteFramerate.Fps25)!.Value;
        var schedule = new MtcSchedule(start, anchor, 1_000_000_000);

        // 25 fps × 4 = 100 messages per second at nanosecond ticks.
        Assert.Equal(10_000_000.0, schedule.TicksPerQuarterFrame);
        Assert.Equal(anchor, schedule.TimestampForQuarterFrame(0));
        Assert.Equal(anchor + 80_000_000, schedule.TimestampForQuarterFrame(8));

        // Negative indices clamp rather than running backwards past the anchor.
        Assert.Equal(anchor, schedule.TimestampForQuarterFrame(-5));
        Assert.Equal(schedule.ByteForQuarterFrame(0), schedule.ByteForQuarterFrame(-5));

        // Eight messages carry one value and advance it by two frames.
        Assert.Equal(start, schedule.TimecodeForSequence(0));
        Assert.Equal(start.FrameCount + 2, schedule.TimecodeForSequence(1).FrameCount);
    }

    /// <summary>
    /// The quantise in <see cref="MtcSchedule"/> is load-bearing, not padding.
    /// Reproduces what a port without it would do and counts the damage, so the
    /// number in the contract's header is an assertion rather than a claim: over
    /// the first 200 messages at nanosecond ticks a tiled chain drops 74 at
    /// 24 fps and 66 at 30 fps, and none at 25 fps where the quarter-frame period
    /// is exactly 10 000 000 ticks.
    /// </summary>
    [Theory]
    [InlineData(SmpteFramerate.Fps24, 74)]
    [InlineData(SmpteFramerate.Fps30, 66)]
    [InlineData(SmpteFramerate.Fps25, 0)]
    public void DroppingTheQuantise_LosesMessages(SmpteFramerate rate, int expectedDropped)
    {
        const ulong anchor = 1000;
        var schedule = new MtcSchedule(
            Timecode.Create(0, 0, 59, 20, rate)!.Value, anchor, 1_000_000_000);
        var period = schedule.TicksPerQuarterFrame;

        var dropped = Enumerable.Range(0, 200).Count(index =>
        {
            var offset = (double)schedule.TimestampForQuarterFrame(index) - anchor;
            if (!(offset > 0))
            {
                return false;
            }

            // No quantise: `Ceiling` of a value a hair above the integer index
            // lands on the *next* index, so this message is never emitted.
            return (int)Math.Ceiling(offset / period) != index;
        });

        Assert.Equal(expectedDropped, dropped);
    }

    /// <summary>
    /// A coarse host clock makes every odd quarter-frame land on a <c>.5</c>
    /// tick, which is the only case that separates Swift's half-away-from-zero
    /// <c>.rounded()</c> from .NET's banker's <c>Math.Round</c> inside
    /// <c>TimestampForQuarterFrame</c>.
    /// </summary>
    [Fact]
    public void CoarseClock_RoundsHalfAwayFromZero()
    {
        const ulong anchor = 1000;
        var schedule = new MtcSchedule(
            Timecode.Create(0, 0, 59, 20, SmpteFramerate.Fps25)!.Value, anchor, 50);

        Assert.Equal(0.5, schedule.TicksPerQuarterFrame);
        Assert.Equal(anchor + 1, schedule.TimestampForQuarterFrame(1));  // 0.5 → 1, not 0
        Assert.Equal(anchor + 3, schedule.TimestampForQuarterFrame(5));  // 2.5 → 3, not 2
        Assert.Equal(0, Math.Round(0.5));                                // banker's, for contrast
        Assert.Equal(2, Math.Round(2.5));
    }

    /// <summary>A host clock that does not advance traps on macOS
    /// (<c>precondition</c>); it must fail here too rather than dividing by
    /// zero and scheduling everything at the anchor.</summary>
    [Fact]
    public void NonPositiveTicksPerSecond_Throws()
    {
        var timecode = Timecode.Create(0, 0, 0, 0, SmpteFramerate.Fps25)!.Value;
        Assert.Throws<ArgumentOutOfRangeException>(() => new MtcSchedule(timecode, 0, 0));
        Assert.Throws<ArgumentOutOfRangeException>(() => new MtcSchedule(timecode, 0, -1_000_000_000));
    }

    /// <summary>The degenerate windows, pinned independently of the vector: an
    /// empty or inverted window yields nothing, and one opening before the anchor
    /// clamps to quarter-frame 0 instead of going negative.</summary>
    [Fact]
    public void DegenerateWindows_YieldNothingOrClampToZero()
    {
        const ulong anchor = 1000;
        var schedule = new MtcSchedule(
            Timecode.Create(0, 0, 59, 20, SmpteFramerate.Fps24)!.Value, anchor, 1_000_000_000);

        Assert.Empty(schedule.Batch(anchor, anchor));
        Assert.Empty(schedule.Batch(anchor + 500, anchor + 100));
        Assert.Empty(schedule.Batch(0, anchor));
        // Half a quarter-frame past the anchor, so the clamped index 0 is inside
        // the window. A tick past it would *not* be: the quantise pulls `until`
        // back onto index 0, and `[from, until)` excludes its own upper bound.
        Assert.Empty(schedule.Batch(0, anchor + 1));
        Assert.Equal(anchor, Assert.Single(schedule.Batch(0, anchor + 5_000_000)).Timestamp);
    }
}
